File: frontend/src/types/exame.ts
import { Antimicrobiano } from "./antimicrobiano";
import { Material } from "./material";
import { Microrganismo } from "./microrganismo";
import { Paciente } from "./paciente";
import { Setor } from "./setor";
import { TipoCultura } from "./tipoCultura";

/**
 * Espelha `app/schemas/exame.py` do backend (fluxo de Exame unificado -
 * Fase 1/2). `Exame` é a entidade "dona" de todo o fluxo: resolve o
 * paciente por prontuário, e grava o pedido + resultado + isolados +
 * antibiograma numa única operação.
 */

export type StatusExame =
  | "AGUARDANDO_TRIAGEM"
  | "NEGATIVO_PARCIAL"
  | "POSITIVO_PARCIAL"
  | "NEGATIVO"
  | "POSITIVO"
  | "CONTAMINACAO";

export const STATUS_EXAME_OPCOES: StatusExame[] = [
  "AGUARDANDO_TRIAGEM",
  "NEGATIVO_PARCIAL",
  "POSITIVO_PARCIAL",
  "NEGATIVO",
  "POSITIVO",
  "CONTAMINACAO",
];

export const STATUS_EXAME_LABELS: Record<StatusExame, string> = {
  AGUARDANDO_TRIAGEM: "Aguardando triagem",
  NEGATIVO_PARCIAL: "Negativo parcial",
  POSITIVO_PARCIAL: "Positivo parcial",
  NEGATIVO: "Negativo",
  POSITIVO: "Positivo",
  CONTAMINACAO: "Contaminação",
};

export const STATUS_EXAME_BADGE_CLASSNAME: Record<StatusExame, string> = {
  AGUARDANDO_TRIAGEM: "mg-badge-info",
  NEGATIVO_PARCIAL: "mg-badge-alerta",
  POSITIVO_PARCIAL: "mg-badge-alerta",
  NEGATIVO: "mg-badge-sucesso",
  POSITIVO: "mg-badge-erro",
  CONTAMINACAO: "mg-badge-erro",
};

// Estados "em andamento" (ainda não finalizados) - espelha
// STATUS_EM_ANDAMENTO do backend (app/models/exame.py).
export const STATUS_EM_ANDAMENTO: StatusExame[] = [
  "AGUARDANDO_TRIAGEM",
  "NEGATIVO_PARCIAL",
  "POSITIVO_PARCIAL",
];

// Status a partir dos quais é possível lançar isolados - espelha
// STATUS_POSITIVO do backend.
export const STATUS_PERMITE_ISOLADOS: StatusExame[] = ["POSITIVO_PARCIAL", "POSITIVO"];

// Status finais aceitos pela ação de "Liberar" na listagem - o único jeito
// de finalizar um exame hoje é um PUT com um desses três valores.
export const STATUS_LIBERACAO_OPCOES: StatusExame[] = ["NEGATIVO", "POSITIVO", "CONTAMINACAO"];

export type MecanismoResistencia =
  | "NENHUM"
  | "MRSA"
  | "ESBL"
  | "CARBAPENEMASE_KPC"
  | "VRE"
  | "D_TESTE_POSITIVO"
  | "OUTRO";

export const MECANISMO_RESISTENCIA_OPCOES: MecanismoResistencia[] = [
  "NENHUM",
  "MRSA",
  "ESBL",
  "CARBAPENEMASE_KPC",
  "VRE",
  "D_TESTE_POSITIVO",
  "OUTRO",
];

export const MECANISMO_RESISTENCIA_LABELS: Record<MecanismoResistencia, string> = {
  NENHUM: "Nenhum",
  MRSA: "MRSA",
  ESBL: "ESBL",
  CARBAPENEMASE_KPC: "Carbapenemase (KPC)",
  VRE: "VRE",
  D_TESTE_POSITIVO: "D-teste positivo",
  OUTRO: "Outro",
};

export type ResultadoSIR = "SENSIVEL" | "INTERMEDIARIO" | "RESISTENTE";

export const RESULTADO_SIR_OPCOES: ResultadoSIR[] = ["SENSIVEL", "INTERMEDIARIO", "RESISTENTE"];

export const RESULTADO_SIR_LABELS: Record<ResultadoSIR, string> = {
  SENSIVEL: "Sensível (S)",
  INTERMEDIARIO: "Intermediário (I)",
  RESISTENTE: "Resistente (R)",
};

export interface ExameAntibiogramaIn {
  antimicrobiano_id: string;
  resultado: ResultadoSIR;
}

export interface ExameAntibiogramaOut {
  antimicrobiano: Antimicrobiano;
  resultado: ResultadoSIR;
}

export interface ExameIsoladoIn {
  microrganismo_id: string;
  mecanismo_resistencia: MecanismoResistencia;
  nao_realizado_tecnico: boolean;
  motivo_dispensa_tsa?: string | null;
  antibiograma: ExameAntibiogramaIn[];
}

export interface ExameIsoladoOut {
  id: string;
  microrganismo: Microrganismo;
  mecanismo_resistencia: MecanismoResistencia;
  nao_realizado_tecnico: boolean;
  motivo_dispensa_tsa: string | null;
  antibiograma: ExameAntibiogramaOut[];
}

export interface ExameCreate {
  paciente_prontuario: string;
  paciente_nome: string;
  setor_id?: string | null;
  tipo_cultura_id: string;
  material_id: string;
  data_coleta?: string | null;
  previsao_liberacao?: string | null;
  status: StatusExame;
  identificacao_preliminar?: string | null;
  observacoes?: string | null;
  isolados: ExameIsoladoIn[];
}

export interface ExameUpdate {
  setor_id?: string | null;
  tipo_cultura_id?: string | null;
  material_id?: string | null;
  data_coleta?: string | null;
  previsao_liberacao?: string | null;
  status?: StatusExame;
  identificacao_preliminar?: string | null;
  observacoes?: string | null;
  isolados?: ExameIsoladoIn[];
}

export interface ExameOut {
  id: string;
  paciente_id: string;
  paciente: Paciente | null;
  setor_id: string | null;
  setor: Setor | null;
  tipo_cultura_id: string;
  tipo_cultura: TipoCultura | null;
  material_id: string;
  material: Material | null;
  data_coleta: string;
  previsao_liberacao: string | null;
  status: StatusExame;
  identificacao_preliminar: string | null;
  observacoes: string | null;
  isolados: ExameIsoladoOut[];
  is_active: boolean;
  created_at: string;
  updated_at: string;
}

export interface ExameListagem {
  total: number;
  page: number;
  page_size: number;
  items: ExameOut[];
}

/**
 * Preview extraído automaticamente de um laudo colado pelo usuário
 * (`POST /api/exames/{id}/importar-laudo`). `*_id` vem `null` quando o
 * nome citado no laudo não bateu exatamente com o catálogo do tenant -
 * nesse caso o usuário precisa escolher manualmente no formulário.
 */
export interface LaudoImportadoAntibiograma {
  antimicrobiano_id: string | null;
  antimicrobiano_nome_laudo: string;
  resultado: ResultadoSIR;
}

export interface LaudoImportadoIsolado {
  microrganismo_id: string | null;
  microrganismo_nome_laudo: string;
  mecanismo_resistencia: MecanismoResistencia;
  antibiograma: LaudoImportadoAntibiograma[];
}

export interface LaudoImportado {
  paciente_confere: boolean;
  prontuario_laudo: string | null;
  nome_paciente_laudo: string | null;
  isolados: LaudoImportadoIsolado[];
}

File: backend/app/schemas/exame.py
"""
Schemas do fluxo de Exame unificado (Fase 1).

`Exame` é agora a entidade "dona" de todo o fluxo (não compõe mais
Solicitação + Cultura) - resolve o paciente por prontuário (cria se não
existir) e grava diretamente o pedido + resultado + isolados +
antibiograma numa única operação. Ver app/services/exame_service.py.
"""
import uuid
from datetime import date, datetime

from pydantic import BaseModel, ConfigDict, Field, model_validator

from app.models.exame import MecanismoResistenciaEnum, ResultadoSIREnum, StatusExameEnum
from app.schemas.antimicrobiano import AntimicrobianoOut
from app.schemas.material import MaterialOut
from app.schemas.microrganismo import MicrorganismoOut
from app.schemas.paciente import PacienteOut
from app.schemas.setor import SetorOut
from app.schemas.tipo_cultura import TipoCulturaOut


class ExameAntibiogramaIn(BaseModel):
    antimicrobiano_id: uuid.UUID
    resultado: ResultadoSIREnum


class ExameIsoladoIn(BaseModel):
    microrganismo_id: uuid.UUID
    mecanismo_resistencia: MecanismoResistenciaEnum = MecanismoResistenciaEnum.NENHUM
    nao_realizado_tecnico: bool = Field(
        default=False,
        description="Dispensa este isolado específico da exigência de "
        "antibiograma (ex.: microrganismo sem protocolo BrCAST).",
    )
    motivo_dispensa_tsa: str | None = Field(default=None, max_length=500)
    antibiograma: list[ExameAntibiogramaIn] = Field(default_factory=list)

    @model_validator(mode="after")
    def _validar_motivo_dispensa(self) -> "ExameIsoladoIn":
        if self.nao_realizado_tecnico and not self.motivo_dispensa_tsa:
            raise ValueError(
                "motivo_dispensa_tsa é obrigatório quando nao_realizado_tecnico=True."
            )
        return self


class ExameCreate(BaseModel):
    paciente_prontuario: str = Field(
        ...,
        min_length=1,
        max_length=50,
        description="Se já existir um paciente com este prontuário, ele é "
        "reaproveitado (o nome cadastrado NÃO é sobrescrito). Caso "
        "contrário, um novo paciente é criado com 'paciente_nome'.",
    )
    paciente_nome: str = Field(
        ...,
        min_length=2,
        max_length=200,
        description="Usado apenas para cadastrar um paciente novo, quando o "
        "prontuário informado ainda não existe.",
    )
    setor_id: uuid.UUID | None = None
    numero_solicitacao: str | None = Field(default=None, max_length=50)
    tipo_cultura_id: uuid.UUID
    material_id: uuid.UUID
    data_coleta: datetime | None = Field(
        default=None,
        description="Se não informada, usa o momento do cadastro (assume que a "
        "coleta acabou de acontecer).",
    )
    previsao_liberacao: date | None = Field(
        default=None,
        description="Se não informado, é calculado automaticamente a partir do "
        "parâmetro 'prazo_solicitacao_dias'.",
    )
    status: StatusExameEnum = StatusExameEnum.AGUARDANDO_TRIAGEM
    identificacao_preliminar: str | None = Field(default=None, max_length=300)
    observacoes: str | None = Field(default=None, max_length=1000)
    isolados: list[ExameIsoladoIn] = Field(default_factory=list)


class ExameUpdate(BaseModel):
    """Todos os campos opcionais - permite atualização parcial (PATCH)."""

    setor_id: uuid.UUID | None = None
    numero_solicitacao: str | None = Field(default=None, max_length=50)
    tipo_cultura_id: uuid.UUID | None = None
    material_id: uuid.UUID | None = None
    data_coleta: datetime | None = None
    previsao_liberacao: date | None = None
    status: StatusExameEnum | None = None
    identificacao_preliminar: str | None = Field(default=None, max_length=300)
    observacoes: str | None = Field(default=None, max_length=1000)
    isolados: list[ExameIsoladoIn] | None = None


class ExameAntibiogramaOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    antimicrobiano: AntimicrobianoOut
    resultado: ResultadoSIREnum


class ExameIsoladoOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    microrganismo: MicrorganismoOut
    mecanismo_resistencia: MecanismoResistenciaEnum
    nao_realizado_tecnico: bool
    motivo_dispensa_tsa: str | None
    antibiograma: list[ExameAntibiogramaOut] = []


class ExameOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    paciente_id: uuid.UUID
    paciente: PacienteOut | None = None
    setor_id: uuid.UUID | None
    setor: SetorOut | None = None
    tipo_cultura_id: uuid.UUID
    tipo_cultura: TipoCulturaOut | None = None
    material_id: uuid.UUID
    material: MaterialOut | None = None
    numero_solicitacao: str | None = None
    data_coleta: datetime
    previsao_liberacao: date | None
    status: StatusExameEnum
    identificacao_preliminar: str | None
    observacoes: str | None
    isolados: list[ExameIsoladoOut] = []
    is_active: bool
    created_at: datetime
    updated_at: datetime


class ExameParcialOut(ExameOut):
    pendencia: str


class ExameListOut(BaseModel):
    total: int
    page: int
    page_size: int
    items: list[ExameOut]


class ExameParcialListOut(BaseModel):
    total: int
    items: list[ExameParcialOut]


class PainelLinhaOut(BaseModel):
    """Linha do Painel de Acompanhamento (ver ExameService.painel_acompanhamento)."""

    id: uuid.UUID
    prontuario: str | None
    paciente_nome: str | None
    numero_solicitacao: str | None
    data_coleta: datetime
    tipo_cultura: str | None
    material: str | None
    setor: str | None
    status: StatusExameEnum
    status_painel: str
    microrganismos: list[str]
    resistencia: list[str]
    sensibilidade: list[str]
    sensivel_exposicao_aumentada: list[str]
    mecanismos_resistencia: list[str]
